### PR TITLE
Feature/initial service

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Initial search API service to handle requests to search for a dimension within a hierarchy of a versioned dataset (instance).

### How to review

Check logic makes sense, all unit test pass and api reflects spec.

### Who can review

Team B

### CODE REVIEW

File: elasticsearch/api.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+	listOfScores := []Scores{}
+	listOfScores = append(listOfScores, scores)
+
+	query := &Body{
+		From: offset,
+		Size: limit,
+		Highlight: &Highlight{
+			PreTags:  []string{"\001S"},
+			PostTags: []string{"\001E"},
```
Comment (gedge) [inline elasticsearch/api.go]: these two lines are inconsistent with other uses of the tags (which use the `"\u0001S"` Unicode forms) - can they be standardised?

File: main.go [hunk 2]
```diff
@@ -1,20 +1,71 @@
...
+
+	createSearchIndexProducer, err := kafka.NewProducer(cfg.Brokers, cfg.HierarchyBuiltTopic, 0)
+	if err != nil {
+		log.Error(errors.Wrap(err, "error creating kakfa producer"), nil)
+		os.Exit(1)
+	}
+
+	datasetAPI := dataset.NewDatasetAPI(client, cfg.DatasetAPIURL)
 
+	// TODO This needs to be sanitised to hide sensitive configs
```
Comment (gedge) [inline main.go]: and you're working on that _right now_, right?

File: service/service.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	healthTicker := healthcheck.NewTicker(
+		svc.HealthCheckInterval,
+		svc.DatasetAPI.GetHealthCheckClient(),
+		elasticsearch.NewHealthCheckClient(svc.ElasticsearchURL),
+	)
+
+	api.CreateSearchAPI(svc.SearchAPIURL, svc.BindAddr, svc.SecretKey, svc.DatasetAPISecretKey, apiErrors, svc.SearchIndexProducer, svc.DatasetAPI, svc.Elasticsearch, svc.DefaultMaxResults)
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
```
Comment (gedge) [inline service/service.go]: this doesn't need to be a function, just put the func's contents after the `for` loop (that is on line 78), at the end of this func

File: service/service.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+		healthTicker.Close()
+
+		log.Info("shutdown complete", nil)
+
+		cancel()
+		os.Exit(1)
+	}
+
+	for {
```
Comment (gedge) [inline service/service.go]: no need for a loop here, just the `select`

File: service/service.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		select {
```
Comment (gedge) [inline service/service.go]: add comment `// blocks until a fatal error occurs` perhaps?

File: auth/authentication_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestMiddleWareAuthenticationReturnsForbidden(t *testing.T) {
+	t.Parallel()
+	Convey("When no access token is provide, unauthorised status code is returned", t, func() {
+		auth := &Authenticator{"123", "internal-token"}
+		r, err := http.NewRequest("POST", "http://localhost:21800/instances", nil)
```
[the remark this answers is not in the document]
Comment (nshumoogum) [reply auth/authentication_test.go]: No it doesn't matter, it's just a place holder for the request path.

File: dataset/api.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+// GetVersion queries the dataset API to get a version
+func (api *API) GetVersion(ctx context.Context, datasetID, edition, version, authToken string) (versionDoc *models.Version, err error) {
+	path := api.url + "/datasets/" + datasetID + "/editions/" + edition + "/versions/" + version
+	logData := log.Data{"func": "GetVersion", "url": path, "dataset_id": datasetID, "edition": edition, "version": version}
+
+	jsonResult, httpCode, err := api.get(ctx, path, authToken, nil)
+	logData["http_code"] = httpCode
+	logData["json_result"] = jsonResult
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (nshumoogum) [reply dataset/api.go]: yes :)

**Pull request merged**